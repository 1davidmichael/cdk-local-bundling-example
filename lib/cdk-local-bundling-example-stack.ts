import * as cdk from '@aws-cdk/core';
import { Code, Runtime, Function} from '@aws-cdk/aws-lambda';
import * as path from 'path';
import { execSync } from 'child_process';

export class CdkLocalBundlingExampleStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const functionDir = path.join(__dirname, "functions", "exampleFunction")

    const exampleFunction = new Function(this, "ExampleFunction", {
      handler: 'index.handler',
      runtime: Runtime.PYTHON_3_8,
      code: Code.fromAsset(functionDir, {
        bundling: {
          image: Runtime.PYTHON_3_8.bundlingImage,
          local: {
            tryBundle(outputDir: string) {
              try {
                execSync('pip3 --version')
              } catch {
                return false
              }

              execSync(`pip install -r ${path.join(functionDir, "requirements.txt")} -t ${path.join(outputDir)}`)
              execSync(`cp -au ${functionDir}/* ${path.join(outputDir)}`)
              return true
            }
          }
        }
      })
    })
  }
}
