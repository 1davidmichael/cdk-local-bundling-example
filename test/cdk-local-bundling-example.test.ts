import { expect as expectCDK, matchTemplate, MatchStyle, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as myApp from '../lib/cdk-local-bundling-example-stack'

test('Empty Stack', () => {
    const app = new cdk.App();
    // WHEN
    const stack = new myApp.CdkLocalBundlingExampleStack(app, 'MyTestStack');
    // THEN
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});