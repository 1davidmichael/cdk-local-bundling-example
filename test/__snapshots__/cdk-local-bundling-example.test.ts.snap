// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty Stack 1`] = `
Object {
  "Parameters": Object {
    "AssetParameterscace00dce7603543ef5e695b67d9df6eeca98bb280af45f9fa920cc5e0e665ceArtifactHash13B21F7B": Object {
      "Description": "Artifact hash for asset \\"cace00dce7603543ef5e695b67d9df6eeca98bb280af45f9fa920cc5e0e665ce\\"",
      "Type": "String",
    },
    "AssetParameterscace00dce7603543ef5e695b67d9df6eeca98bb280af45f9fa920cc5e0e665ceS3Bucket8BB26296": Object {
      "Description": "S3 bucket for asset \\"cace00dce7603543ef5e695b67d9df6eeca98bb280af45f9fa920cc5e0e665ce\\"",
      "Type": "String",
    },
    "AssetParameterscace00dce7603543ef5e695b67d9df6eeca98bb280af45f9fa920cc5e0e665ceS3VersionKey16A2A4C7": Object {
      "Description": "S3 key for asset version \\"cace00dce7603543ef5e695b67d9df6eeca98bb280af45f9fa920cc5e0e665ce\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ExampleFunctionB28997EC": Object {
      "DependsOn": Array [
        "ExampleFunctionServiceRoleA44A0F8B",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameterscace00dce7603543ef5e695b67d9df6eeca98bb280af45f9fa920cc5e0e665ceS3Bucket8BB26296",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterscace00dce7603543ef5e695b67d9df6eeca98bb280af45f9fa920cc5e0e665ceS3VersionKey16A2A4C7",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterscace00dce7603543ef5e695b67d9df6eeca98bb280af45f9fa920cc5e0e665ceS3VersionKey16A2A4C7",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ExampleFunctionServiceRoleA44A0F8B",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ExampleFunctionServiceRoleA44A0F8B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
